pkg/operator/kuryr.go

```diff
@@ -0,0 +1,146 @@
+package operator
+
+import (
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	netv1 "github.com/openshift/api/network/v1"
+	"github.com/operator-framework/operator-sdk/pkg/util/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/cluster-network-operator/pkg/apis/networkoperator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/render"
+)
+
+// renderKuryr returns manifests for Kuryr SDN.
+// This creates
+// - the ClusterNetwork object
+// - the kuryr namespace
+// - the kuryr-controller deployment
+// - the kuryr-daemon daemonset
+// and some other small things.
+func (h *Handler) renderKuryr(c *v1.KuryrConfig) ([]*uns.Unstructured, error) {
+	operConfig := h.config.Spec
+	objs := []*uns.Unstructured{}
+
+	// generate master network configuration
+	ippools := []netv1.ClusterNetworkEntry{}
+	for _, net := range operConfig.ClusterNetworks {
+		ippools = append(ippools, netv1.ClusterNetworkEntry{CIDR: net.CIDR, HostSubnetLength: net.HostSubnetLength})
+	}
+
+	clusterNet := netv1.ClusterNetwork{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "network.openshift.io/v1",
+			Kind:       "ClusterNetwork",
+		},
+		ObjectMeta: metav1.ObjectMeta{Name: netv1.ClusterNetworkDefault},
+
+		ServiceNetwork:  operConfig.ServiceNetwork,
+		PluginName:      sdnPluginName("OpenStack Kuryr-Kubernetes"),
+		ClusterNetworks: ippools,
+		// VXLANPort:       c.VXLANPort,
+
+		Network:          ippools[0].CIDR,
+		HostSubnetLength: ippools[0].HostSubnetLength,
+	}
+	obj, err := k8sutil.UnstructuredFromRuntimeObject(&clusterNet)
+	if err != nil {
+		// This is very unlikely
+		return nil, errors.Wrap(err, "failed to transmutate ClusterNetwork")
+	}
+	objs = append(objs, obj)
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+
+	// OpenStack credentials
+	if c.OpenStackCACertificate == nil {
+		data.Data["OpenStackCACertificate"] = ""
+	} else {
+		data.Data["OpenStackCACertificate"] = *c.OpenStackCACertificate
+	}
+	data.Data["OpenStackAuthURL"] = *c.OpenStackCredentials.AuthURL
+	data.Data["OpenStackPassword"] = *c.OpenStackCredentials.Password
+	data.Data["OpenStackProjectDomainName"] = *c.OpenStackCredentials.ProjectDomainName
+	data.Data["OpenStackProjectId"] = *c.OpenStackCredentials.ProjectId
+	data.Data["OpenStackUserDomainName"] = *c.OpenStackCredentials.UserDomainName
+	data.Data["OpenStackUsername"] = *c.OpenStackCredentials.Username
+	data.Data["KuryrProject"] = c.KuryrProject
+
+	// TODO(dulek): Figure out how do I get those
+	data.Data["PodSubnet"] = ""
+	data.Data["PodSecurityGroups"] = ""
+	data.Data["ServiceSubnet"] = ""
+	data.Data["WorkerNodesSubnet"] = ""
```

Comment:
I was thinking that since they will be created by terraform in openshift/installer, that probably openshift/installer should create a ConfigMap asset that we mount also on /etc/kuryr/ and leverage oslo config functionality of reading multiple config files.

@squeed @dcbw Would that be an installer + operator idiomatic flow for the SDN needing things created during deployment?

Reply:
(obviously we would only mount the configmap on the controller, I forgot to specify :P )

Reply:
Another way could be to make openshift/installer put that onto the NetworkConfig object its (I assume?) creating. See the manifests/0007_10_cluster-network-operator_01_crd.yaml file. I'm assuming whole user-facing Kuryr configuration is read from there.